### ISSUE TITLE
Error extracting from in memory zip file

### ISSUE THREAD

**Issue by nedway** [gtreshchev/RuntimeArchiver#4]
```
[2024.01.25-14.38.45:392][359]LogRuntimeArchiver: Error: ERuntimeArchiverErrorCode::ExtractError: Unable to extract zip entry 'SaveGameInfoJSON.txt' into memory.
Miniz error details: 'invalid header or archive is corrupted'.
```

the issue is this:
i am using 
```
	UFUNCTION(BlueprintCallable, Category = "Runtime Archiver|Open")
	bool OpenArchiveFromMemory(TArray<uint8> ArchiveData);
```

but there are several problems with this function as written:
1 - it makes a temporary copy of the archive data
2 - it makes a _second_ temporary copy in 
```
bool URuntimeArchiverBase::OpenArchiveFromMemory(TArray<uint8> ArchiveData)
{
	return OpenArchiveFromMemory(TArray64<uint8>(MoveTemp(ArchiveData)));
}
```
_that_ copy gets destructed on the way out of OpenArchiveFromMemory, which ends up freeing the memory for the temporary copy and overwriting the first four bytes of the archive data

that makes the first file unreadable, of course ( it appears corrupted because it is missing a proper signature ) but of course the problem is worse than that since the whole block has been free and a raw pointer to it stored in MinizArchiver->m_pState.m_pMem

I can hack arounds this by changing 

```
	UFUNCTION(BlueprintCallable, Category = "Runtime Archiver|Open")
	bool OpenArchiveFromMemory(const TArray<uint8> &ArchiveData);
```
and
```
bool URuntimeArchiverBase::OpenArchiveFromMemory(const TArray<uint8> &ArchiveData)
{
	const TArrayView64<const uint8> ArchiveData64( (const uint8 *) ArchiveData.GetData(), ArchiveData.Num());

	return OpenArchiveFromMemory(*reinterpret_cast<const TArray64<uint8>*>(&ArchiveData64));
}

```
but *that* is kind of a rickety abomination

**Comment by gtreshchev** [comment]
I think there may be some misunderstanding: 

1. It doesn't create any copies during the `MoveTemp` call, when passing `TArray<uint8>` to a function that gets `TArray64<uint8>`, because it uses the move constructor to pass the array from the 32-bit-specific function to the 64-bit one. I recommend you look into the `static void MoveOrCopy(ToArrayType& ToArray, FromArrayType& FromArray, SizeType PrevMax)` templated function to see how it works. Essentially, if it's possible to move pointers between arrays, they're easily movable without any extra cost (you can manually check that the `UE4Array_Private::CanMoveTArrayPointersBetweenArrayTypes<TArray<uint8>, TArray64<uint8>>()` returns true).

2. In case you are not using the `MoveTemp` on the function that receives `TArray<uint8>`, it creates just a single copy, when passing the binary data to the 32-bit-specific `OpenArchiveFromMemory`, but you can avoid even that copy if you pass your binary array through `MoveTemp`. There will be zero reallocations in this case, and it will always use your passed binary data within your array, within the same pointer. 

Regarding `overwriting the first four bytes of the archive data`, I don't think that array does any of it internally. A good way to debug that would be to get the array size of `TArray<uint8>` and compare it with the `TArray64<uint8>` one, and there won't be any difference in that.

**Comment by gtreshchev** [comment]
I think there may be some misunderstanding: 

1. It doesn't create any copies during the `MoveTemp` call, when passing `TArray<uint8>` to a function that gets `TArray64<uint8>`, because it uses the move constructor to pass the array from the 32-bit-specific function to the 64-bit one. I recommend you look into the `static void MoveOrCopy(ToArrayType& ToArray, FromArrayType& FromArray, SizeType PrevMax)` templated function to see how it works. Essentially, if it's possible to move pointers between arrays, they're easily movable without any extra cost (you can manually check that the `UE4Array_Private::CanMoveTArrayPointersBetweenArrayTypes<TArray<uint8>, TArray64<uint8>>()` returns true).

2. In case you are not using the `MoveTemp` on the function that receives `TArray<uint8>`, it creates just a single copy, when passing the binary data to the 32-bit-specific `OpenArchiveFromMemory`, but you can avoid even that copy if you pass your binary array through `MoveTemp`. There will be zero reallocations in this case, and it will always use your passed binary data within the same pointer. 

Regarding `overwriting the first four bytes of the archive data`, I don't think that array does any of it internally. A good way to debug that would be to get the array size of `TArray<uint8>` and compare it with the `TArray64<uint8>` one, and there won't be any difference in that.

**Comment by gtreshchev** [comment]
I've just looked through the code and it turns out there isn't a function to pass `TArray64<uint8>` by value, it's only available through a const reference. So I think I'll need to create a separate function that assumes passing `TArray64<uint8>` by r-value reference, rather than just by const l-value reference. You can do it manually for now.

**Comment by gtreshchev** [comment]
Should be fixed in 0bbf1a78afac7852236841f679967f98adcb09f4.

**Comment by gtreshchev** [comment]
@nedway Please feel free to reopen if it's still relevant or if you have any other context.
